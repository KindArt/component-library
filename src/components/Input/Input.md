Standard:

```jsx
<Input label="Label" placeholder="Placeholder" />
```

With Icon:

```jsx
const icon = <div>✊🏾</div>;
<Input icon={icon} iconPosition="left" />;
```

Disabled:

```jsx padded
<Input label="Label" placeholder="Placeholder" disabled />
```

Error State:

```jsx
<Input label="Label" errorMessage="There is an error" />
```
