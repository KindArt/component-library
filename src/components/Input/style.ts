import styled, { css } from 'styled-components';
import { HTMLProps } from 'react';
import { Colours, CssVariables } from '../../constants/styles';
import { IInputProps } from '.';

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;

  ${({ className }) =>
    className &&
    css`
      ${className}
    `}
`;

export const InputWrapper = styled.div`
  display: flex;
  position: relative;
`;

interface IIconWrapper {
  iconPosition: 'left' | 'right';
}

export const IconWrapper = styled.span<IIconWrapper>`
  position: absolute;
  top: 50%;
  color: ${({ theme }) => theme.input.iconWrapperColour};
  transform: translateY(-50%);

  ${({ iconPosition }) =>
    iconPosition === 'left' &&
    css`
      left: 12px;
      & + ${InputField} {
        padding-left: 40px;
      }
    `}

  ${({ iconPosition }) =>
    iconPosition === 'right' &&
    css`
      // left: auto;
      right: 12px;
      & + ${InputField} {
        padding-right: 40px;
      }
    `}

  ${({ className }) =>
    className &&
    css`
      ${className}
    `}
`;

IconWrapper.defaultProps = {
  theme: {
    input: {
      iconWrapperColour: Colours.BORDER_COLOUR_DARKER,
    },
  },
};

export const InputField = styled.input<IInputProps>`
  border: ${({ theme }) => theme.input.border};
  border-radius: 4px;
  flex: 1;
  padding: 10px 15px;
  font-size: 0.875rem;

  &:disabled {
    background: rgba(62, 63, 66, 0.3);
    cursor: not-allowed;
  }

  &:focus {
    border-color: ${({ theme }) => theme.input.focusColour};
  }

  ${({ errorMessage, theme }) =>
    errorMessage &&
    css`
      border-color: ${theme.input.errorBorderColor};
    `}

  ${({ className }) =>
    className &&
    css`
      ${className}
    `}
`;

InputField.defaultProps = {
  theme: {
    input: {
      border: `${CssVariables.BORDER_DEFAULT}`,
      errorBorderColor: `${Colours.ERROR}`,
      focusColour: `${Colours.PRIMARY}`,
    },
  },
};

export const ErrorMessage = styled.p`
  color: ${({ theme }) => theme.errorMessage.color};
  margin-top: 6px;
  margin-bottom: 0;
  line-height: 1.375rem;

  ${({ className }) =>
    className &&
    css`
      ${className}
    `}
`;

ErrorMessage.defaultProps = {
  theme: {
    errorMessage: {
      color: `${Colours.ERROR}`,
    },
  },
};
