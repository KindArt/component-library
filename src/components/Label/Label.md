Standard:

```jsx
<Label content="This is the content" />
```
