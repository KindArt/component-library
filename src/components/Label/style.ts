import styled, { css } from 'styled-components';
import { Colours } from '../../constants/styles';

export const LabelWrapper = styled.span`
  font-weight: 500;
  font-size: 0.75rem;
  color: ${({ theme }) => theme.label.colour};
  line-height: 1.125rem;
  text-transform: uppercase;
  & + * {
    margin-top: 0.5rem;
  }

  ${({ className }) =>
    className &&
    css`
      ${className}
    `}
`;

LabelWrapper.defaultProps = {
  theme: {
    label: {
      colour: `${Colours.TEXT_SECONDARY}`,
    },
  },
};
