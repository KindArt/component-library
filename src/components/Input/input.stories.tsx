import React from 'react';
import Input from '.';

export default {
  component: Input,
  title: 'Input',
  decorators: [(storyFn: any) => <div style={{ maxWidth: '50%' }}>{storyFn()}</div>]
};

const iconMock = (
  <span role="img" aria-label="">
    ✊🏾
  </span>
);

export const normal = () => <Input />;

export const placeholder = () => <Input placeholder="Placeholder" />;

export const errorState = () => <Input errorMessage="There is an error!" />;

export const withLabel = () => <Input label="Label" />;

export const LHSIcon = () => <Input icon={iconMock} />;

export const RHSIcon = () => <Input icon={iconMock} iconPosition="right" />;
