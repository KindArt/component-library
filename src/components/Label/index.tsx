import React, { FC, HTMLProps } from 'react';
import { LabelWrapper } from './style';

interface ILabel extends HTMLProps<HTMLLabelElement> {
  id?: string;
  content: string;
  testIds?: {
    wrapper?: string;
  };
  classOverrides?: {
    wrapper?: string;
  };
}

const componentName = 'Label';

const Label: FC<ILabel> = ({ id, content, testIds, classOverrides, ...props }) => {
  const tidPrefix = id || componentName;
  const tid = {
    wrapper: (testIds && testIds.wrapper) || 'wrapper',
  };

  return (
    <LabelWrapper
      className={classOverrides && classOverrides.wrapper}
      data-testid={`${tidPrefix}-${tid.wrapper}`}
      {...(props as any)}
    >
      {content}
    </LabelWrapper>
  );
};

export default Label;
