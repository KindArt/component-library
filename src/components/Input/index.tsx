import React, { HTMLProps, ReactNode, forwardRef } from 'react';
import { FormGroupWrapper, ErrorMessage, IconWrapper, InputWrapper, InputField } from './style';
import Label from '../Label';

enum ElementIdentifiers {
  formGroup = 'formGroup',
  label = 'label',
  icon = 'icon',
  input = 'input',
  errorMessage = 'errorMessage',
}

export interface IInputProps extends HTMLProps<HTMLInputElement> {
  label?: string;
  icon?: ReactNode;
  iconPosition?: 'left' | 'right';
  errorMessage?: string;
  value?: string;
  testIds?: {
    formGroup?: string;
    label?: string;
    icon?: string;
    input?: string;
    errorMessage?: string;
  };
}

const componentName = 'Input';

const Input = forwardRef<HTMLInputElement, IInputProps>(
  ({ id, label, icon, iconPosition = 'left', errorMessage, value, testIds, type = 'text', ...props }, ref) => {
    const tidPrefix = id || componentName;
    const tid = {
      formGroup: (testIds && testIds.formGroup) || ElementIdentifiers.formGroup,
      label: (testIds && testIds.label) || ElementIdentifiers.label,
      icon: (testIds && testIds.icon) || ElementIdentifiers.icon,
      input: (testIds && testIds.input) || ElementIdentifiers.input,
      errorMessage: (testIds && testIds.errorMessage) || ElementIdentifiers.errorMessage,
    };
    return (
      <FormGroupWrapper data-testId={`${tidPrefix}-${tid.formGroup}`}>
        {label && <Label data-testId={`${tidPrefix}-${tid.label}`} content={label} />}
        <InputWrapper>
          {icon && (
            <IconWrapper data-testId={`${tidPrefix}-${tid.icon}`} iconPosition={iconPosition}>
              {icon}
            </IconWrapper>
          )}
          <InputField
            ref={ref}
            data-testId={`${tidPrefix}-${tid.input}`}
            type={type}
            errorMessage={errorMessage}
            value={value}
            {...(props as any)}
          ></InputField>
        </InputWrapper>
        {errorMessage && <ErrorMessage data-testId={`${tidPrefix}-${tid.errorMessage}`}>{errorMessage}</ErrorMessage>}
      </FormGroupWrapper>
    );
  }
);

export default Input;
