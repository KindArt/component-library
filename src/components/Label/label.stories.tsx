import React from 'react';
import Label from '.';

export default {
  component: Label,
  title: 'Label',
};

export const normal = () => <Label content="This is a label" />;
