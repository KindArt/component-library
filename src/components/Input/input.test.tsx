import React from 'react';
import { shallow } from 'enzyme';
import Label from '../Label';
import Input from '.';

describe('<Input />', () => {
  let wrapper: any;
  beforeEach(() => {
    wrapper = shallow(<Input />);
  });

  it('should render', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('Visuals', () => {
    it('should display a label if one is passed', () => {
      wrapper.setProps({ label: 'Label' });
      const label = wrapper.find(Label);
      const actual = label.prop('content');
      const expected = 'Label';
      expect(actual).toEqual(expected);
    });

    it('should display an icon if one is passed', () => {
      wrapper.setProps({ icon: <div></div> });
      const actual = wrapper.find('[data-testId="Input-icon"]').length;
      const expected = 1;
      expect(actual).toEqual(expected);
    });

    it('should display an error message if one is passed', () => {
      wrapper.setProps({ errorMessage: 'Error' });
      const actual = wrapper.find('[data-testId="Input-errorMessage"]').length;
      const expected = 1;
      expect(actual).toEqual(expected);
    });

    it('should default icon placement to the left', () => {
      wrapper.setProps({ icon: <div></div> });
      const icon = wrapper.find('[data-testId="Input-icon"]');
      const actual = icon.prop('iconPosition');
      const expected = 'left';
      expect(actual).toEqual(expected);
    });
  });

  describe('Test Ids', () => {
    it('should set a custom test id for wrapper', () => {
      wrapper.setProps({ testIds: { wrapper: 'customWrapper' } });
      const actual = wrapper.find('[data-testId="Input-customWrapper"]').length;
      const expected = 1;
      expect(actual).toEqual(expected);
    });

    it('should set a custom test id for icon', () => {
      wrapper.setProps({
        testIds: { icon: 'customIcon' },
        icon: <div></div>,
      });
      const actual = wrapper.find('[data-testId="Input-customIcon"]').length;
      const expected = 1;
      expect(actual).toEqual(expected);
    });

    it('should set a custom test id for input', () => {
      wrapper.setProps({ testIds: { input: 'customInput' } });
      const actual = wrapper.find('[data-testId="Input-customInput"]').length;
      const expected = 1;
      expect(actual).toEqual(expected);
    });

    it('should set a custom test id for error', () => {
      wrapper.setProps({
        testIds: { errorMessage: 'customError' },
        errorMessage: 'Error',
      });
      const actual = wrapper.find('[data-testId="Input-customError"]').length;
      const expected = 1;
      expect(actual).toEqual(expected);
    });
  });

  describe('Class Overrides', () => {});
});
