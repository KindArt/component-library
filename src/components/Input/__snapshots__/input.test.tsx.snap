// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should render 1`] = `
<styled.div
  data-testId="Input-formGroup"
>
  <styled.div>
    <styled.input
      data-testId="Input-input"
      theme={
        Object {
          "input": Object {
            "border": "1px solid #EAEDF3",
            "errorBorderColor": "#E6492D",
          },
        }
      }
      type="text"
    />
  </styled.div>
</styled.div>
`;
